src/besselj.jl

```diff
@@ -3,32 +3,98 @@
 #
 #    Calculation of besselj0 is done in three branches using polynomial approximations
 #
-#    Branch 1: x <= 5.0
-#              besselj0 = (x^2 - r1^2)*(x^2 - r2^2)*P3(x^2) / Q8(x^2)
-#    where r1 and r2 are zeros of J0
-#    and P3 and Q8 are a 3 and 8 degree polynomial respectively
-#    Polynomial coefficients are from [1] which is based on [2]
-#    For tiny arugments the power series expansion is used.
+#    Branch 1: x <= pi/2
+#              besselj0 is calculated using a 9 term, even minimax polynomial
 #
-#    Branch 2: 5.0 < x < 25.0
-#              besselj0 = sqrt(2/(pi*x))*(cos(x - pi/4)*R7(x) - sin(x - pi/4)*R8(x))
-#    Hankel's asymptotic expansion is used
-#    where R7 and R8 are rational functions (Pn(x)/Qn(x)) of degree 7 and 8 respectively
-#    See section 4 of [3] for more details and [1] for coefficients of polynomials
-# 
-#   Branch 3: x >= 25.0
+#    Branch 2: pi/2 < x < 26.0
+#              besselj0 is calculated by one of 16 different degree 13 minimax polynomials
+#       Each polynomial is an expansion around either a root or extrema of the besselj0.
+#       This ensures accuracy near the roots. Method taken from [2]
+#
+#   Branch 3: x >= 26.0
 #              besselj0 = sqrt(2/(pi*x))*beta(x)*(cos(x - pi/4 - alpha(x))
-#   See modified expansions given in [3]. Exact coefficients are used
+#   See modified expansions given in [2]. Exact coefficients are used.
 #
 #   Calculation of besselj1 is done in a similar way as besselj0.
-#   See [3] for details on similarities.
-# 
+#   See [2] for details on similarities.
+#
 # [1] https://github.com/deepmind/torch-cephes
-# [2] Cephes Math Library Release 2.8:  June, 2000 by Stephen L. Moshier
-# [3] Harrison, John. "Fast and accurate Bessel function computation." 
+# [2] Harrison, John. "Fast and accurate Bessel function computation."
 #     2009 19th IEEE Symposium on Computer Arithmetic. IEEE, 2009.
 #
 
+J0_ROOTS(::Type{Float64}) = (
+        (2.4048255576957730, -1.176691651530894e-16),
+        (3.8317059702075125, -1.5269184090088067e-16),
+        (5.5200781102863110, 8.088597146146722e-17),
+        (7.0155866698156190, -9.414165653410389e-17),
+        (8.6537279129110130, -2.92812607320779e-16),
+        (10.173468135062722, 4.482162274768888e-16),
+        (11.791534439014281, 2.812956912778735e-16),
+        (13.323691936314223, 2.600408064718813e-16),
+        (14.930917708487787, -7.070514505983074e-16),
+        (16.470630050877634, -1.619019544798128e-15),
+        (18.071063967910924, -9.658048089426209e-16),
+        (19.615858510468243, -1.004445634526616e-15),
+        (21.211636629879260, 4.947077428784068e-16),
+        (22.760084380592772, -4.925749373614922e-16),
+        (24.352471530749302, 9.169067133951066e-16),
+        (25.903672087618382, 4.894530726419825e-16)
+)
+@inline J0_POLYS(::Type{Float64}) = (
```

Comment:
Is there a reason we sometimes inline these, sometimes we don't, and sometimes we use constants for the polynomials? It might be good to have a defined way going forward that is consistent. Though this one is slightly different as we have to determine which polynomial at runtime. Does this affect compile time?

Reply:
Also, I would be a fan of moving the polynomials to their own file and slightly different file structure to make the core code easier to read. I think this approach could also be used for `Double64` implementations that would also take up a lot of space in main file.